sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -4600,3 +4600,92 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
 }
+
+/**
+ * Given an array, and another element append the element at the end of the array.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(array, element) - Append the element",
+  examples =
+    """
+    Examples:
+      > SELECT _FUNC_(array('b', 'd', 'c', 'a'), 'd');
+       ["b","d","c","a","d"]
+
+  """,
+  since = "3.4.0",
+  group = "array_funcs")
+case class ArrayAppend(left: Expression, right: Expression)
+  extends BinaryExpression
+  with ImplicitCastInputTypes
+  with ComplexTypeMergingExpression
+  with QueryErrorsBase {
+  override def prettyName: String = "array_append"
+
+  override def inputTypes: Seq[AbstractDataType] = {
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, hasNull), e2) =>
+        TypeCoercion.findTightestCommonType(e1, e2) match {
+          case Some(dt) => Seq(ArrayType(dt, hasNull), dt)
+          case _ => Seq.empty
+        }
+      case _ => Seq.empty
+    }
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, _), e2) if e1.sameType(e2) => TypeCheckResult.TypeCheckSuccess
+      case (ArrayType(e1, _), e2) => DataTypeMismatch(
+          errorSubClass = "ARRAY_FUNCTION_DIFF_TYPES",
+          messageParameters = Map(
+            "functionName" -> toSQLId(prettyName),
+            "leftType" -> toSQLType(left.dataType),
+            "rightType" -> toSQLType(right.dataType),
+            "dataType" -> toSQLType(ArrayType)
+          ))
+      case _ =>
+        DataTypeMismatch(
+          errorSubClass = "UNEXPECTED_INPUT_TYPE",
+          messageParameters = Map(
+            "paramIndex" -> "0",
+            "requiredType" -> toSQLType(ArrayType),
+            "inputSql" -> toSQLExpr(left),
+            "inputType" -> toSQLType(left.dataType)
+          )
+        )
+    }
+  }
+
+  protected def withNewChildrenInternal(newLeft: Expression, newRight: Expression): ArrayAppend =
+    copy(left = newLeft, right = newRight)
+
+  override protected def nullSafeEval(input1: Any, input2: Any): Any = {
+    val arrayData = input1.asInstanceOf[ArrayData]
+    val arrayElementType = dataType.asInstanceOf[ArrayType].elementType
+    val elementData = input2
+    val numberOfElements = arrayData.numElements() + 1
+    if (numberOfElements > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+      throw QueryExecutionErrors.concatArraysWithElementsExceedLimitError(numberOfElements)
+    }
+    val finalData = new Array[Any](numberOfElements)
+    arrayData.foreach(arrayElementType, finalData.update)
+    finalData.update(numberOfElements - 1, elementData)
+    new GenericArrayData(finalData)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    nullSafeCodeGen(
+      ctx, ev, (left: String, right: String) => {
+        val expr = ctx.addReferenceObj("arraysAppendExpr", this)
+        s"${ev.value} = (ArrayData)$expr.nullSafeEval($left, $right);"
```

Comment:
Could you implement `doGenCode` by codegen? Please refer `ArrayRemove`.

Reply:
Personally, the earlier implementation is better.

Reply:
oh is it ? It is actually similar piece of code, I used it since I was overriding the behaviour we had for checking null entries but now since we are kind of NullIntolerant, so this is actually okay since by default it puts null checks cleanly.

Reply:
@beliefer I found that the prs of these 4 new functions all choice this implementation way. What are the advantages of  implement `doGenCode` by codegen over the current implementation way? Thanks !!!

Reply:
I think there are two ways:
1. Implement `doGenCode` by codegen like `ArrayRemove`. it will help to reduce the stack depth.
2. Construct a new companion Object and extract the common implementation as a function of Object. It will help to reuse the common code in `nullSafeEval` and `doGenCode`.

Reply:
Thanks for your answer, which has benefited me a lot @beliefer

Reply:
Thanks for putting these points forward @beliefer ! Let me work on it.

Reply:
@beliefer Updated the code for generating the code for the same manually.

Reply:
@beliefer @LuciferYang I am thinking ArrayAppend and ArrayPrepend can probably be taken care by same code piece, since most of the checks and processings are going to be same. I can create a trait have everything there being controlled by a boolean flag. Let me update the code real quick for the same, I think it will be make both functions cleaner.